modules/core/src/test/java/org/apache/ignite/internal/processors/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,426 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.ignite.IgniteLock;
+import org.apache.ignite.cache.CacheMode;
+import org.apache.ignite.configuration.AtomicConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.util.lang.GridAbsPredicate;
+import org.apache.ignite.spi.discovery.tcp.TcpDiscoverySpi;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.TcpDiscoveryIpFinder;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.vm.TcpDiscoveryVmIpFinder;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+
+/**
+ * Tests for {@link org.apache.ignite.internal.processors.datastructures.GridCacheLockImpl2Fair} and for {@link
+ * org.apache.ignite.internal.processors.datastructures.GridCacheLockImpl2Unfair}
+ */
+public class IgniteReentrantLockTest extends GridCommonAbstractTest {
+    /** IP finder. */
+    private static final TcpDiscoveryIpFinder ipFinder = new TcpDiscoveryVmIpFinder(true);
+
+    /** */
+    private static final int NODES_CNT = 4;
+
+    /** */
+    private static final Random RND = new Random();
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        TcpDiscoverySpi spi = new TcpDiscoverySpi();
+
+        spi.setIpFinder(ipFinder);
+
+        cfg.setDiscoverySpi(spi);
+
+        AtomicConfiguration atomicCfg = new AtomicConfiguration();
+
+        assertNotNull(atomicCfg);
+
+        atomicCfg.setCacheMode(CacheMode.REPLICATED);
+
+        cfg.setAtomicConfiguration(atomicCfg);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        startGridsMultiThreaded(NODES_CNT);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    /**
+     * Test a creating of the reentrant lock.
+     *
+     * @throws Exception If failed.
+     */
+    public void testInitialization() throws Exception {
+        try (IgniteLock lock = createReentrantLock(0, "lock1", false)) {
+        }
+
+        try (IgniteLock lock = createReentrantLock(0, "lock2", true)) {
+        }
+    }
+
+    /**
+     * Test a sequential lock acquiring and releasing.
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock() throws Exception {
+        testReentrantLock(true);
+        testReentrantLock(false);
+    }
+
+    /**
+     * {@link this#testReentrantLock()}
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock(boolean fair) throws Exception {
+        for (int i = 0; i < NODES_CNT; i++) {
+            final IgniteLock lock = createReentrantLock(i, fair ? "lock1" : "lock2", fair);
```

Comment:
mb, "lock1" -> "unfair lock"
"lock2" -> "fair lock" here and there

Reply:
Don't see why

Reply:
"unfair lock" sounds better than "lock1", isn't it ?

Reply:
fine